source/elements/oneMKL/source/domains/spblas/operations/spmm.rst

```diff
@@ -0,0 +1,393 @@
+       Produces deterministic results.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG2``
+       | rocSPARSE: ``rocsparse_spmm_alg_coo_atomic``
+   * - ``coo_alg3``
+     - Should provide best performance for COO format and large ``nnz``.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG3``
+       | rocSPARSE: ``rocsparse_spmm_alg_coo_segmented_atomic``
+   * - ``coo_alg4``
+     - Should provide best performance for COO format and row-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_COO_ALG4``
+       | rocSPARSE: N/A
+   * - ``csr_alg1``
+     - Should provide best performance for CSR format and column-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG1``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr``
+   * - ``csr_alg2``
+     - Should provide best performance for CSR format and row-major layout.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG2``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr_row_split``
+   * - ``csr_alg3``
+     - Deterministic algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMM_CSR_ALG3``
+       | rocSPARSE: ``rocsparse_spmm_alg_csr_merge``
+
+.. _onemkl_sparse_spmm:
+
+spmm
+----
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void spmm_buffer_size(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           oneapi::mkl::transpose                     opB,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_matrix_handle_t B_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+           oneapi::mkl::sparse::spmm_alg              alg,
+           oneapi::mkl::sparse::spmm_descr_t          spmm_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmm_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           oneapi::mkl::transpose                     opB,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_matrix_handle_t B_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_matrix_handle_t C_handle,
+           oneapi::mkl::sparse::spmm_alg              alg,
+           oneapi::mkl::sparse::spmm_descr_t          spmm_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
```

Comment:
I've not done it now to be consistent with the rest of oneMKL but I would also like to see this. I agree it makes more sense:

- `sycl::event` can still be useful with buffers, in particular for getting profiling information from the event.
- This makes the API consistent with USM and makes it easier to switch between the 2.

@gajanan-choudhary and @spencerpatty any thoughts on this before I make this change?

Reply:
Many APIs split their operation into multiple `sycl::event`s instead of doing the entire operation in a single event. The `sycl::event` being returned by the backend may not necessarily be the actual GPU kernel. In some cases, it might just be a `host_task` that's doing nothing inside.

Does profiling an event give profiling info of all the events in the dependency chain before it?
* If not, then the profiling information you get from a returned `sycl::event` from a backend may be incomplete, misleading, and at times even useless.
* If yes, users can just create a `host_task`/`single_task` after calling the oneMKL API with relevant `read` accessors to get profiling info of the dependency chain.

(I believe the answer for the above question is "no", please correct me if I'm wrong.)

Therefore, we'd need to return a `std::vector<sycl::event>` from the APIs. Only backends using SYCL would return meaningful events (meaning cuSPARSE won't, in my understanding). How would you choose which of the events to profile when you don't know what the event is supposed to be doing in the backend?

Overall, I believe having the oneAPI spec reliably enable profiling backend APIs may not be as straightforward.

Reply:
I think you're right about the profiling, it could be complicated to get this to work properly with cuSPARSE with the current SYCL spec. We have found useful to compare the total time spend on the host and the device for benchmarking purposes when possible.

To expand on my previous point on making the 2 APIs consistent, this can be useful:
* internally, since the only difference will be between a USM pointer and a buffer or an accessor, templates can be used to reduce the duplication. We often do this in our portFFT, portBLAS and such.
* for users that need to switch between the 2 or support both containers, for the same reason using templates.

I think this is worth considering. Asking the question the other way around, are there any drawbacks for taking and returning events for the buffer API? I don't expect much performance cost, if any, and we already expect slightly lower performance for buffers anyway.

Reply:
We discussed this issue. There were some concerns with this change.
There isn't that much value to make this change. I'd say we can keep it as it is so we can merge the PR.